File: protocol-generator/dart/bin/peer.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:entity_core_protocol/entity_core_peer.dart';

/// Standalone S4-ready host: boots a peer on a localhost port and prints a
/// `LISTENING <port>` line so a harness can scrape the bound port. Flags:
///
/// ```
///   --port N               bind port (0 = auto, the default)
///   --seed B               seed byte (repeated 32×) for a deterministic identity
///   --name NAME            load a persistent Ed25519 identity from the standard
///                          on-disk location ~/.entity/peers/NAME/keypair (the
///                          entity-core PEM keypair: base64 of a 32-byte seed
///                          between BEGIN/END ENTITY PRIVATE KEY lines — the Go
///                          entity-peer --name / peer-manager convention). Lets the
///                          validator's multisig accept-path probe co-sign AS the
///                          peer (crypto.LookupKeypairByPeerID).
///   --debug-open-grants    degenerate [default → *] admin seed (non-conformant)
///   --validate             bootstrap the §7a system/validate/* conformance handlers
/// ```
///
/// The §7a handlers are OFF by default (a standing dispatch-outbound originator
/// must never ship live); `--validate` opts in (the keystone cohort mechanism).
/// AOT-compiled for S4 (`dart compile exe bin/peer.dart`) per profile [build].
Future<void> main(List<String> args) async {
  var port = 0;
  var seedByte = 1;
  String? name;
  var openGrants = false;
  var validate = false;
  var i = 0;
  while (i < args.length) {
    switch (args[i]) {
      case '--port':
        port = int.parse(args[++i]);
      case '--seed':
        seedByte = int.parse(args[++i]);
      case '--name':
        name = args[++i];
      case '--debug-open-grants':
        openGrants = true;
      case '--validate':
        validate = true;
      default:
        break; // ignore unknown flags for forward-compat
    }
    i++;
  }
  final seed = name != null
      ? _loadSeedFromName(name)
      : (Uint8List(32)..fillRange(0, 32, seedByte));
  final peer =
      await Peer.create(seed, openGrants: openGrants, conformance: validate);
  final listener = await startListener(peer, port);
  stdout.writeln('LISTENING ${listener.port}');
  stdout.writeln('PEER ${peer.localPeer}');
  await stdout.flush();
  // park forever (the harness kills the process)
  await Completer<void>().future;
}

/// Load the 32-byte Ed25519 seed from the standard on-disk keypair at
/// `~/.entity/peers/NAME/keypair` — an entity-core PEM whose body is base64(seed)
/// between BEGIN/END ENTITY PRIVATE KEY lines. Missing or malformed → exit 2.
Uint8List _loadSeedFromName(String name) {
  final home = _firstNonEmpty([
    Platform.environment['HOME'],
    Platform.environment['USERPROFILE'],
  ]) ?? '/root';
  final path = '$home/.entity/peers/$name/keypair';
  try {
    final lines = File(path).readAsLinesSync();
    final body = lines
        .where((l) => !l.startsWith('-'))
        .map((l) => l.trim())
        .join();
    final seed = Uint8List.fromList(base64.decode(body));
    if (seed.length != 32) {
      stderr.writeln(
          'error: --name $name: expected a 32-byte seed, got ${seed.length} bytes');
      exit(2);
    }
    return seed;
  } catch (e) {
    stderr.writeln('error: --name $name: $e');
    exit(2);
  }
}

String? _firstNonEmpty(List<String?> xs) {
  for (final x in xs) {
    if (x != null && x.isNotEmpty) return x;
  }
  return null;
}

File: protocol-generator/julia/bin/peer.jl
#!/usr/bin/env julia
# entity-core-protocol-julia peer executable — the process `validate-peer` drives in S4.
#
# Startup convention (keystone AGENTS.md): `--name NAME` loads the Ed25519 identity from
# ~/.entity/peers/NAME/keypair (entity-core PEM = base64 of a 32-byte seed) → persistent identity
# + peer-manager interop. `--validate` enables the §7a system/validate/* conformance handlers
# (OFF by default — dispatch-outbound is a standing dialer, never live in production).
# `--debug-open-grants` is the deprecated degenerate seed policy (accepted, ignored).
#
# Prints a single `LISTENING 127.0.0.1:PORT` line once the socket is bound (the S4 harness greps
# for it), then serves accepted connections on the single-threaded Task scheduler until killed.
#
# NOTE (S3 boundary): the peer machinery here — handshake, auth-before-resolve dispatch, seed
# grant, single-link root-cap check — is the S3 surface. Full validate-peer conformance (the
# core handlers system/{tree,capability,type,handler}, multi-link chain-walk, §7a validate
# handlers) is the S4 build-out; this entry point is the wiring S4 fills in.
using Sockets
using Base64

const HERE = @__DIR__
include(joinpath(HERE, "..", "src", "EntityCore.jl"))
using .EntityCore
using .EntityCore.Transport: serve_connection

function load_seed(name::AbstractString)::Vector{UInt8}
    home = get(ENV, "HOME", "/root")
    path = joinpath(home, ".entity", "peers", name, "keypair")
    isfile(path) || error("keypair not found: $path (provision it per the --name convention)")
    b64 = ""
    for line in eachline(path)
        startswith(line, "-----") && continue
        isempty(strip(line)) && continue
        b64 = strip(line); break
    end
    seed = base64decode(b64)
    length(seed) == 32 || error("keypair seed must be 32 bytes, got $(length(seed))")
    return seed
end

function main(args)
    port = 7777
    name = "conformance"
    validate = false
    open_grants = false
    i = 1
    while i <= length(args)
        a = args[i]
        if a == "--port"; port = parse(Int, args[i+1]); i += 2
        elseif a == "--name"; name = args[i+1]; i += 2
        elseif a == "--validate"; validate = true; i += 1
        elseif a == "--debug-open-grants"; open_grants = true; i += 1  # degenerate seed policy default→*
        else; i += 1
        end
    end

    peer = create_peer(load_seed(name); validate=validate, open_grants=open_grants)
    server = Sockets.listen(ip"127.0.0.1", port)
    bound = Int(getsockname(server)[2])
    println("LISTENING 127.0.0.1:$bound")
    flush(stdout)
    while true
        sock = try; accept(server); catch; break; end
        @async serve_connection(peer, sock)
    end
end

main(ARGS)
